uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js

```diff
@@ -341,24 +354,32 @@ define([
        * Also, checks if its an update request from manager and accordingly updates leave request
        */
       this.submit = function () {
```

Comment:
I think the whole method could be streamlined as such

```js
this.submit = function () {
  var originalStatus = this.request.status_id;

  if (this.isMode('view') || this.submitting) {
    return;
  }

  validate()
    .then(function () {
      this.submitting = true;

      changeStatusBeforeSave.call(this);
    })
    .then(function () {
      return /* updateRequest() or createRequest() */
    })
    .catch(function (errors) {
      this.request.status_id = originalStatus;

      errors && handleError.call(this, errors);
    })
    .finally(function () {
      this.submitting = false;
    }.bind(this))
};
```
This to me reads much better and it's easier to reason about. Following is the annotated code

```js
this.submit = function () {
  var originalStatus = this.request.status_id;

  // this was in `validateBeforeSubmit` but this is not validation logic
  // it simply a "gate" to prevent the accidental submit when the user is not supposed to
  if (this.isMode('view') || this.submitting) {
    return;
  }

  // this will contain both the current content of `validateBeforeSubmit` but also a call to `this.request.isValid()`
  // the objective is to encapsulate all the validation logic in just one method. The method will reject a promise if 
  // some validation will fail
  validate()
    .then(function () {
      // validation is successful, begin submitting
      this.submitting = true;

      // change the status
      changeStatusBeforeSave.call(this);
    })
    .then(function () {
      // either update or save
      return /* updateRequest() or createRequest() */
    })
    .catch(function (errors) {
      // the code will go here either from `validate()` or from the update/create request
     
      // in any case, put back the original status
      this.request.status_id = originalStatus;

      // if there are any errors, handle them
      errors && handleError.call(this, errors);
    })
    .finally(function () {
      // in any case, remove the submitting state
      this.submitting = false;
    }.bind(this))
};
```